circuit/types/integers/src/div_wrapped.rs

```diff
@@ -63,6 +63,7 @@ impl<E: Environment, I: IntegerType> Integer<E, I> {
     /// Divides `self` by `other`, via witnesses, returning the quotient and remainder.
     /// This method does not check that `other` is non-zero.
     /// This method should only be used when 2 * I::BITS < E::BaseField::size_in_data_bits().
+    /// This method assumes the `self` and `other` are unsigned integers.
```

Comment:
In the current usage of `unsigned_division_via_witness`, these conditions are satisfied or left intentionally unchecked.
If the preconditions are unsatisfied, the circuit would still be able to built but may be incorrect.

The first condition is not explicitly enforced because sometimes we do not want to perform it (when we know that it is always satisfied, like in `shr_wrapped`)
The other two could be checked, though the third may require some modification of the traits.
And we could always add debug assertions throughout the codebase.

Broadly speaking, we would benefit from an audit of our circuit libraries and a following discussion on user-safety.

Reply:
Yeah that is all understandable. Trait modification is less than simple and often opens up its own vulnerabilities in cryptographic applications. I think given its not a public method it *should* be safe - especially when the developers using it are familiar with the assumptions behind it. Perhaps we just add it to an audit tracking issue.